src/topic.ts

```diff
@@ -717,8 +717,100 @@ export class Topic {
     const attributes = typeof attrsOrCb === 'object' ? attrsOrCb : {};
     callback = typeof attrsOrCb === 'function' ? attrsOrCb : callback;
 
-    const data = Buffer.from(JSON.stringify(json));
-    return this.publish(data, attributes, callback!);
+    return this.publishMessage({json, attributes}, callback!);
+  }
+
+  publishMessage(message: MessageOptions): Promise<[string]>;
+  publishMessage(message: MessageOptions, callback: PublishCallback): void;
+  /**
+   * @typedef {object} MessageOptions
+   * @property {buffer} [data] The message data.
+   * @property {object} [json] Convenience property to publish JSON data. This
+   *     will transform the provided JSON into a Buffer before publishing.
+   *     {@link Subscription} objects will always return message data in the
+   *     form of a Buffer, so any JSON published will require manual
+   *     deserialization.
+   * @property {object.<string, string>} [attributes] Attributes for this
+   *     message.
+   * @property {string} [orderingKey] A message ordering key.
+   */
+  /**
+   * Publish the provided message.
+   *
+   * @throws {TypeError} If data is not a Buffer object.
+   * @throws {TypeError} If any value in `attributes` object is not a string.
+   *
+   * @param {MessageOptions} message Message object.
+   * @param {PublishCallback} [callback] Callback function.
+   * @returns {Promise<PublishResponse>}
+   *
+   * @example
+   * const {PubSub} = require('@google-cloud/pubsub');
+   * const pubsub = new PubSub();
+   * const topic = pubsub.topic('my-topic');
+   *
+   * const data = Buffer.from('Hello, world!');
+   *
+   * const callback = (err, messageId) => {
+   *   if (err) {
+   *     // Error handling omitted.
+   *   }
+   * };
+   *
+   * topic.publishMessage({data}, callback);
+   *
+   * @example <caption>Publish JSON message data.</caption>
+   * const json = {foo: 'bar'};
+   *
+   * topic.publishMessage({json}, callback);
+   *
+   * @example <caption>To publish messages in order, make sure message ordering is enabled and provide an ordering key</caption>
```

Comment:
Can we also mention that this feature is currently experimental? We use this language in the Java library: "Ordering is not yet fully supported and requires special project enablements."